File: paleo_workbench/workflow/sample_normalization.py
"""Factor sample normalization — one policy authority for every method (V8 M3).

Twin wells / re-sampled logs deliver samples at IDENTICAL (x, y). Every
interpolation backend treated them differently before V8:

* plain IDW — every duplicate cast a full-weight vote (double-vote);
* constrained IDW — first-wins (its own internal dedupe);
* engine kriging — collapsed to the mean (its own internal dedupe);
* cross-validation — collapsed duplicates for fold scoring while production
  kept them, so CV scored a *different* model than the one delivered.

This module makes the policy EXPLICIT and host-side, applied once before the
engine sees the samples, so production, CV, fingerprints and surface checks
all consume the same normalized sample set:

* ``"mean"`` (default) — duplicates merge to their mean value, matching the
  kriging authority's convention and the CV scorer's existing behaviour;
* ``"first"`` — keep the first occurrence (legacy constrained-IDW semantics);
* ``"error"`` — refuse (strict callers that want duplicates surfaced, never
  silently merged);
* ``"keep"`` — legacy no-op (documented double-vote for plain IDW; escape
  hatch for exact bit-compatibility with pre-V8 results).

The report travels with the task (``parameters["sample_normalization"]``),
the grid result provenance and quality metrics — a merged twin well can never
disappear silently.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any, Mapping, Sequence

__all__ = [
    "DUPLICATE_POLICIES",
    "DEFAULT_DUPLICATE_POLICY",
    "SampleNormalizationReport",
    "normalize_factor_samples",
]


DUPLICATE_POLICIES = ("mean", "first", "error", "keep")
DEFAULT_DUPLICATE_POLICY = "mean"

_NUMERIC_EXTRAS = ("q", "b_i")


@dataclass(frozen=True, slots=True)
class SampleNormalizationReport:
    """Honest accounting of what normalization did to a sample set."""

    policy: str
    n_input: int
    n_valid: int
    n_nonfinite_dropped: int
    n_duplicate_groups: int  # distinct (x, y) locations with 2+ samples
    n_duplicates_merged: int  # total samples removed by merging
    n_qc_flagged: int  # samples carrying a non-ok qc_flag (visibility only)

    @property
    def duplicates_present(self) -> bool:
        return self.n_duplicate_groups > 0

    def to_dict(self) -> dict[str, Any]:
        return {
            "policy": self.policy,
            "n_input": self.n_input,
            "n_valid": self.n_valid,
            "n_nonfinite_dropped": self.n_nonfinite_dropped,
            "n_duplicate_groups": self.n_duplicate_groups,
            "n_duplicates_merged": self.n_duplicates_merged,
            "n_qc_flagged": self.n_qc_flagged,
        }


def _valid_record(pt: Any) -> dict[str, Any] | None:
    """Valid (finite x, y, z) sample record, or None (mirrors the
    fingerprint extractor's validity rules so both authorities agree)."""
    if not isinstance(pt, Mapping):
        return None
    try:
        if "x" in pt and "y" in pt:
            x = float(pt["x"])
            y = float(pt["y"])
        elif "lng" in pt and "lat" in pt:
            x = float(pt["lng"])
            y = float(pt["lat"])
        else:
            return None
        z = float(pt.get("value", pt.get("z", pt.get("v"))))
    except (TypeError, ValueError):
        return None
    if not (math.isfinite(x) and math.isfinite(y) and math.isfinite(z)):
        return None
    record: dict[str, Any] = {"x": x, "y": y, "z": z}
    for key in _NUMERIC_EXTRAS:
        value = pt.get(key)
        if value is not None:
            try:
                number = float(value)
            except (TypeError, ValueError):
                continue
            if math.isfinite(number):
                record[key] = number
    qc = pt.get("qc_flag")
    if qc is not None:
        record["qc_flag"] = str(qc)
    for key in ("well_id", "name"):
        value = pt.get(key)
        if value:
            record[key] = str(value)
    return record


def normalize_factor_samples(
    points: Sequence[Any],
    *,
    policy: str = DEFAULT_DUPLICATE_POLICY,
) -> tuple[list[dict[str, Any]], SampleNormalizationReport]:
    """Apply the duplicate policy to a raw sample set.

    Returns ``(normalized_points, report)``. Non-finite samples are dropped
    (the engine would drop them anyway — here the count is reported instead
    of being invisible). Order is preserved: the first occurrence of each
    distinct location defines the output position of its merged record.
    """
    if policy not in DUPLICATE_POLICIES:
        raise ValueError(
            f"unknown duplicate policy {policy!r}; choose from {DUPLICATE_POLICIES}"
        )
    records: list[dict[str, Any]] = []
    n_nonfinite = 0
    for pt in points or []:
        record = _valid_record(pt)
        if record is None:
            n_nonfinite += 1
            continue
        records.append(record)
    n_qc_flagged = sum(
        1 for r in records if str(r.get("qc_flag") or "") not in ("", "ok")
    )

    if policy == "keep":
        report = SampleNormalizationReport(
            policy=policy,
            n_input=len(points or []),
            n_valid=len(records),
            n_nonfinite_dropped=n_nonfinite,
            n_duplicate_groups=0,
            n_duplicates_merged=0,
            n_qc_flagged=n_qc_flagged,
        )
        return records, report

    # Group by exact coordinate identity (engine dedupe semantics: exact
    # equality, no epsilon — near-but-not-equal wells stay distinct samples).
    groups: dict[tuple[float, float], list[int]] = {}
    order: list[tuple[float, float]] = []
    for index, record in enumerate(records):
        key = (record["x"], record["y"])
        if key not in groups:
            groups[key] = []
            order.append(key)
        groups[key].append(index)

    duplicate_keys = [key for key in order if len(groups[key]) > 1]
    if policy == "error" and duplicate_keys:
        raise ValueError(
            "duplicate sample locations present and duplicate_policy='error': "
            f"{len(duplicate_keys)} location(s) with 2+ samples "
            f"(first at x={duplicate_keys[0][0]!r}, y={duplicate_keys[0][1]!r})"
        )

    normalized: list[dict[str, Any]] = []
    merged = 0
    for key in order:
        indices = groups[key]
        if len(indices) == 1:
            normalized.append(records[indices[0]])
            continue
        merged += len(indices) - 1
        members = [records[i] for i in indices]
        base = dict(members[0])
        if policy == "mean":
            # incremental mean: a plain sum() overflows to inf for huge
            # values (1e308 duplicates); mean-of-two must stay finite
            def _imean(values):
                acc = 0.0
                for i, v in enumerate(values, 1):
                    acc += (v - acc) / i
                return acc

            base["z"] = _imean(m["z"] for m in members)
            for extra in _NUMERIC_EXTRAS:
                values = [m[extra] for m in members if extra in m]
                if values:
                    base[extra] = _imean(values)
        # Worst-flag-wins: one flagged twin flags the merged location — a
        # conservative sample is dropped by QC-aware backends, never laundered.
        flags = {str(m.get("qc_flag") or "") for m in members}
        flags.discard("")
        flags.discard("ok")
        if flags:
            base["qc_flag"] = sorted(flags)[0]
        names = [str(m.get("well_id") or m.get("name") or "") for m in members]
        names = [n for n in names if n]
        if names:
            base["well_id"] = "+".join(dict.fromkeys(names))
        normalized.append(base)

    report = SampleNormalizationReport(
        policy=policy,
        n_input=len(points or []),
        n_valid=len(records),
        n_nonfinite_dropped=n_nonfinite,
        n_duplicate_groups=len(duplicate_keys),
        n_duplicates_merged=merged,
        n_qc_flagged=n_qc_flagged,
    )
    return normalized, report


def duplicate_policy_from_params(params: Mapping[str, Any] | None) -> str:
    """Resolve the task's duplicate policy with honest fallback to default."""
    raw = (params or {}).get("duplicate_policy")
    if raw is None:
        return DEFAULT_DUPLICATE_POLICY
    policy = str(raw).strip().lower()
    if policy not in DUPLICATE_POLICIES:
        return DEFAULT_DUPLICATE_POLICY
    return policy

File: libs/mapping_kernel/src/sample_normalization.cpp
#include <pwb/mapping/sample_normalization.hpp>

#include <algorithm>
#include <cctype>
#include <cmath>
#include <cstdlib>
#include <limits>
#include <map>
#include <optional>
#include <sstream>
#include <stdexcept>
#include <string>
#include <unordered_set>
#include <utility>
#include <vector>

namespace pwb::mapping {
namespace {

bool known_policy(const std::string& policy) {
    for (const char* p : kDuplicatePolicies) {
        if (policy == p) return true;
    }
    return false;
}

std::string py_repr_float(double v) {
    std::ostringstream o;
    o.setf(std::ios::fmtflags(0), std::ios::floatfield);
    o << v;
    std::string s = o.str();
    if (s.find('.') == std::string::npos && s.find('e') == std::string::npos
        && s.find('E') == std::string::npos) {
        s += ".0";
    }
    return s;
}

bool json_to_float(const Json& v, double& out) {
    if (v.is_number()) {
        out = v.get<double>();
        return true;
    }
    if (v.is_boolean()) {
        out = v.get<bool>() ? 1.0 : 0.0;
        return true;
    }
    if (!v.is_string()) return false;
    const std::string s = v.get<std::string>();
    if (s == "inf" || s == "Infinity" || s == "+inf") {
        out = std::numeric_limits<double>::infinity();
        return true;
    }
    if (s == "-inf" || s == "-Infinity") {
        out = -std::numeric_limits<double>::infinity();
        return true;
    }
    char* end = nullptr;
    out = std::strtod(s.c_str(), &end);
    return end != s.c_str() && end != nullptr && *end == '\0';
}

bool json_truthy(const Json& v) {
    if (v.is_null()) return false;
    if (v.is_boolean()) return v.get<bool>();
    if (v.is_string()) return !v.get<std::string>().empty();
    if (v.is_number()) return v.get<double>() != 0.0;
    if (v.is_array() || v.is_object()) return !v.empty();
    return true;
}

std::string json_str(const Json& v) {
    if (v.is_string()) return v.get<std::string>();
    if (v.is_boolean()) return v.get<bool>() ? "True" : "False";
    if (v.is_number_unsigned()) return std::to_string(v.get<std::uint64_t>());
    if (v.is_number_integer()) return std::to_string(v.get<std::int64_t>());
    if (v.is_number_float()) return Json(v.get<double>()).dump();
    if (v.is_null()) return "None";
    return v.dump();
}

std::optional<Json> valid_record(const Json& pt) {
    if (!pt.is_object()) return std::nullopt;
    double x = 0.0, y = 0.0, z = 0.0;
    if (pt.contains("x") && pt.contains("y")) {
        if (!json_to_float(pt["x"], x) || !json_to_float(pt["y"], y)) {
            return std::nullopt;
        }
    } else if (pt.contains("lng") && pt.contains("lat")) {
        if (!json_to_float(pt["lng"], x) || !json_to_float(pt["lat"], y)) {
            return std::nullopt;
        }
    } else {
        return std::nullopt;
    }
    Json zsrc = Json();
    if (pt.contains("value")) zsrc = pt["value"];
    else if (pt.contains("z")) zsrc = pt["z"];
    else if (pt.contains("v")) zsrc = pt["v"];
    else return std::nullopt;
    if (!json_to_float(zsrc, z)) return std::nullopt;
    if (!(std::isfinite(x) && std::isfinite(y) && std::isfinite(z))) {
        return std::nullopt;
    }
    Json rec = Json::object();
    rec["x"] = x;
    rec["y"] = y;
    rec["z"] = z;
    for (const char* key : {"q", "b_i"}) {
        if (!pt.contains(key) || pt[key].is_null()) continue;
        double n = 0.0;
        if (json_to_float(pt[key], n) && std::isfinite(n)) rec[key] = n;
    }
    if (pt.contains("qc_flag") && !pt["qc_flag"].is_null()) {
        rec["qc_flag"] = json_str(pt["qc_flag"]);
    }
    for (const char* key : {"well_id", "name"}) {
        if (!pt.contains(key)) continue;
        if (json_truthy(pt[key])) rec[key] = json_str(pt[key]);
    }
    return rec;
}

double incremental_mean(const std::vector<double>& values) {
    double acc = 0.0;
    int i = 0;
    for (double v : values) {
        ++i;
        acc += (v - acc) / static_cast<double>(i);
    }
    return acc;
}

}  // namespace

std::string duplicate_policy_from_params(const Json& params) {
    if (!params.is_object() || !params.contains("duplicate_policy")
        || params["duplicate_policy"].is_null()) {
        return kDefaultDuplicatePolicy;
    }
    std::string policy = json_str(params["duplicate_policy"]);
    // strip + lower
    while (!policy.empty() && std::isspace(static_cast<unsigned char>(policy.front()))) {
        policy.erase(policy.begin());
    }
    while (!policy.empty() && std::isspace(static_cast<unsigned char>(policy.back()))) {
        policy.pop_back();
    }
    for (char& c : policy) {
        c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
    }
    if (!known_policy(policy)) return kDefaultDuplicatePolicy;
    return policy;
}

std::pair<Json, SampleNormalizationReport> normalize_factor_samples(
    const Json& points, const std::string& policy) {
    if (!known_policy(policy)) {
        throw std::invalid_argument(
            "unknown duplicate policy '" + policy
            + "'; choose from ('mean', 'first', 'error', 'keep')");
    }
    const Json arr = points.is_array() ? points : Json::array();
    std::vector<Json> records;
    int n_nonfinite = 0;
    for (const auto& pt : arr) {
        auto rec = valid_record(pt);
        if (!rec.has_value()) {
            ++n_nonfinite;
            continue;
        }
        records.push_back(*rec);
    }
    int n_qc = 0;
    for (const auto& r : records) {
        const std::string flag =
            r.contains("qc_flag") ? r["qc_flag"].get<std::string>() : "";
        if (flag != "" && flag != "ok") ++n_qc;
    }

    SampleNormalizationReport report;
    report.policy = policy;
    report.n_input = static_cast<int>(arr.size());
    report.n_valid = static_cast<int>(records.size());
    report.n_nonfinite_dropped = n_nonfinite;
    report.n_qc_flagged = n_qc;

    if (policy == "keep") {
        Json out = Json::array();
        for (const auto& r : records) out.push_back(r);
        return {out, report};
    }

    std::map<std::pair<double, double>, std::vector<int>> groups;
    std::vector<std::pair<double, double>> order;
    for (int i = 0; i < static_cast<int>(records.size()); ++i) {
        const auto key = std::make_pair(records[static_cast<std::size_t>(i)]["x"].get<double>(),
                                        records[static_cast<std::size_t>(i)]["y"].get<double>());
        if (groups.find(key) == groups.end()) order.push_back(key);
        groups[key].push_back(i);
    }
    std::vector<std::pair<double, double>> duplicate_keys;
    for (const auto& key : order) {
        if (groups[key].size() > 1) duplicate_keys.push_back(key);
    }
    if (policy == "error" && !duplicate_keys.empty()) {
        throw std::invalid_argument(
            "duplicate sample locations present and duplicate_policy='error': "
            + std::to_string(duplicate_keys.size())
            + " location(s) with 2+ samples (first at x="
            + py_repr_float(duplicate_keys[0].first) + ", y="
            + py_repr_float(duplicate_keys[0].second) + ")");
    }

    Json normalized = Json::array();
    int merged = 0;
    for (const auto& key : order) {
        const auto& indices = groups[key];
        if (indices.size() == 1) {
            normalized.push_back(records[static_cast<std::size_t>(indices[0])]);
            continue;
        }
        merged += static_cast<int>(indices.size()) - 1;
        Json base = records[static_cast<std::size_t>(indices[0])];
        if (policy == "mean") {
            std::vector<double> zs;
            for (int idx : indices) {
                zs.push_back(records[static_cast<std::size_t>(idx)]["z"].get<double>());
            }
            base["z"] = incremental_mean(zs);
            for (const char* extra : {"q", "b_i"}) {
                std::vector<double> vals;
                for (int idx : indices) {
                    const Json& m = records[static_cast<std::size_t>(idx)];
                    if (m.contains(extra)) vals.push_back(m[extra].get<double>());
                }
                if (!vals.empty()) base[extra] = incremental_mean(vals);
            }
        }
        std::vector<std::string> flags;
        for (int idx : indices) {
            const Json& m = records[static_cast<std::size_t>(idx)];
            std::string f = m.contains("qc_flag") ? m["qc_flag"].get<std::string>() : "";
            if (f != "" && f != "ok") flags.push_back(f);
        }
        if (!flags.empty()) {
            std::sort(flags.begin(), flags.end());
            // unique? Python set then sorted — unique values, first of sorted
            flags.erase(std::unique(flags.begin(), flags.end()), flags.end());
            base["qc_flag"] = flags.front();
        }
        std::vector<std::string> names;
        std::unordered_set<std::string> seen;
        for (int idx : indices) {
            const Json& m = records[static_cast<std::size_t>(idx)];
            std::string n;
            if (m.contains("well_id")) n = m["well_id"].get<std::string>();
            else if (m.contains("name")) n = m["name"].get<std::string>();
            if (n.empty() || seen.count(n)) continue;
            seen.insert(n);
            names.push_back(n);
        }
        if (!names.empty()) {
            std::string joined = names[0];
            for (std::size_t i = 1; i < names.size(); ++i) {
                joined += "+";
                joined += names[i];
            }
            base["well_id"] = joined;
        }
        normalized.push_back(base);
    }
    report.n_duplicate_groups = static_cast<int>(duplicate_keys.size());
    report.n_duplicates_merged = merged;
    return {normalized, report};
}

}  // namespace pwb::mapping
